### PR TITLE
update PodSpec.Host to PodSpec.NodeName

### LINKED ISSUE #6895: Rename PodSpec.Host to PodSpec.NodeName
It's minor, but now that I see it, I can't unsee it.

@bgrant0607

### PR DESCRIPTION
**Pull request by caesarxuchao**
update PodSpec.Host to PodSpec.NodeName in /pkg/api/types.go and /pkg/api/v1/types.go

fix  #6895 

@bgrant0607 @krousey

### CODE REVIEW

File: pkg/api/v1/types.go [hunk 1]
```diff
@@ -845,10 +845,10 @@ type PodSpec struct {
 	// ServiceAccount is the name of the ServiceAccount to use to run this pod
 	ServiceAccount string `json:"serviceAccount,omitempty" description:"name of the ServiceAccount to use to run this pod"`
 
-	// Host is a request to schedule this pod onto a specific host.  If it is non-empty,
+	// NodeName is a request to schedule this pod onto a specific host.  If it is non-empty,
 	// the the scheduler simply schedules this pod onto that host, assuming that it fits
 	// resource requirements.
-	Host string `json:"host,omitempty" description:"host requested for this pod"`
+	NodeName string `json:"nodeName,omitempty" description:"host requested for this pod"`
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Please change the description tag, also: node requested for this pod.

File: pkg/kubectl/describe.go [hunk 2]
```diff
@@ -269,7 +269,7 @@ func describePod(pod *api.Pod, rcs []api.ReplicationController, events *api.Even
 	return tabbedString(func(out io.Writer) error {
 		fmt.Fprintf(out, "Name:\t%s\n", pod.Name)
 		fmt.Fprintf(out, "Image(s):\t%s\n", makeImageList(&pod.Spec))
-		fmt.Fprintf(out, "Host:\t%s\n", pod.Spec.Host+"/"+pod.Status.HostIP)
+		fmt.Fprintf(out, "Host:\t%s\n", pod.Spec.NodeName+"/"+pod.Status.HostIP)
```
Comment (bgrant0607) [inline pkg/kubectl/describe.go]: Change the row title to Node, also.

**Comment by bgrant0607** [comment]
Fails gofmt

**Comment by bgrant0607** [comment]
Ref #7018

File: pkg/kubectl/describe.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/kubectl/describe.go]: I will update pther places with similar problem. I was not sure which way was better so I didn't update them.

Comment (bgrant0607) [inline pkg/kubectl/describe.go]: Thanks. We'll likely change this, anyway (see #7843), but let's go with Node.

**Comment by caesarxuchao** [comment]
Updated. PTAL. Thanks.

**Comment by bgrant0607** [comment]
Doesn't build. Regenerate pkg/api/deep_copy_generated.go

**Comment by caesarxuchao** [comment]
Yeah, I have the problem when I work on #6797 today. It's strange this problem didn't occur yesterday. I will get back to this PR after fixing #6797.

**Comment by bgrant0607** [comment]
@wojtek-t We're having trouble with the generated conversions. Is it obvious what we're doing wrong?

**Comment by caesarxuchao** [comment]
Hi @wojtek-t, no big deals, I just want to know what's the best practice when one tried to update the api, e.g., update PodSpec.Host to PodSpec.Nodename.

here are the steps that I want to take, but these don't work:
1. change a field name (e.g. Host->NodeName) in api/types.go and api/v1/types.go
2. add functions in v1beta3/conversion.go to convert "Host" in v1beta3 to "NodeName" in internal version
3. run hack/update-generated-conversion.sh and hack/update-generated-deep-copies.sh.
This doesn't work because in step 3, the scripts complain that the v1beta3/conversion-generated.go and deep-copies_generated.go contain undeclared field name "Host"

So I have to manually update various *-generated.go, and then these update-* scripts are happy. However, at this point, I have already made all the changes manually, there is no work left for these update-* scripts. Please let me know if there is a better way. Thanks.

File: pkg/controller/controller_utils.go [hunk 3]
```diff
@@ -19,6 +19,8 @@ package controller
 import (
 	"fmt"
 
+	"sync/atomic"
```
Comment (bgrant0607) [inline pkg/controller/controller_utils.go]: Why this change?

Comment (caesarxuchao) [inline pkg/controller/controller_utils.go]: There is something wrong with the goimports. I will fix it. Thanks.

**Comment by bgrant0607** [comment]
Building the generator should not build the previously generated code. That seems wrong.

**Comment by caesarxuchao** [comment]
In cmd/genconversion/conversion.go:
```
     "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
     _ "github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1"
     _ "github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta1"
     _ "github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta2"
     _ "github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta3"
```
So to build the generator, it requires the conversion_generated.go files in these packages to be correct.

**Comment by bgrant0607** [comment]
I understand that's the way it is. I was saying it shouldn't be that way. :-)

The generated code just needs to build, though, right? It doesn't need to work? If you commented out the lines it complained about, would the generator build and then regenerate the conversions?

**Comment by caesarxuchao** [comment]
Yeah, that should work. That seems to be the best option.

**Comment by wojtek-t** [comment]
@caesarxuchao - yes, unfortunately this is the easiest way to do it. Basically, what you can do is to remove the whole code from */conversions_generated.go files, and running:
hack/update-generated-conversions.sh will regenerate them.
I agree it's not convenient and shouldn't work like this but it's not trivial to fix it.

**Comment by bgrant0607** [comment]
@wojtek-t Please add that to the instructions.
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/api_changes.md#edit-version-conversions

**Comment by wojtek-t** [comment]
@bgrant0607 - I thought about it and I know how to fix the problem. Will fix the real issue instead of updating documentation :)

**Comment by caesarxuchao** [comment]
"remove the whole code from */conversions_generated.go" won't work, because some functions in conversion.go depend on code in conversions_generated.go

@wojtek-t, I'm looking forward to your fix : )

File: pkg/api/validation/validation.go [hunk 4]
```diff
@@ -1003,8 +1003,8 @@ func ValidatePodStatusUpdate(newPod, oldPod *api.Pod) errs.ValidationErrorList {
 	allErrs = append(allErrs, ValidateObjectMetaUpdate(&oldPod.ObjectMeta, &newPod.ObjectMeta).Prefix("metadata")...)
 
 	// TODO: allow change when bindings are properly decoupled from pods
-	if newPod.Spec.Host != oldPod.Spec.Host {
-		allErrs = append(allErrs, errs.NewFieldInvalid("status.host", newPod.Spec.Host, "pod host cannot be changed directly"))
+	if newPod.Spec.NodeName != oldPod.Spec.NodeName {
+		allErrs = append(allErrs, errs.NewFieldInvalid("status.nodeName", newPod.Spec.NodeName, "pod host cannot be changed directly"))
```
Comment (bgrant0607) [inline pkg/api/validation/validation.go]: s/host/node/ in error message

**Comment by wojtek-t** [comment]
@caesarxuchao unfortunately my fix doesn't work because what you wrote. Updated instructions are in #8921

**Comment by bgrant0607** [comment]
Tests failing:
```
_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/deep_copy_generated.go:1377: out.Host undefined (type *PodSpec has no field or method Host)
```

**Comment by caesarxuchao** [comment]
@bgrant0607  Updated and rebased. Let's wait for the Shippable results. Thanks.

**Comment by bgrant0607** [comment]
Shippable failed already:
```
 github.com/GoogleCloudPlatform/kubernetes/test/soak/serve_hostnames
test/soak/serve_hostnames/serve_hostnames.go:188: unknown "github.com/GoogleCloudPlatform/kubernetes/pkg/api".PodSpec field 'Host' in struct literal
```

**Comment by caesarxuchao** [comment]
There is something wrong in field.selector. I'm taking a look.

**Comment by caesarxuchao** [comment]
@bgrant0607 Shippable is green

**Comment by bgrant0607** [comment]
Please squash the commits.

**Comment by bgrant0607** [comment]
LGTM once squashed.

**Comment by caesarxuchao** [comment]
Squashed. Thanks.
